### PR TITLE
rgw: Replace boost::string_ref/view with std::string_view

### PR DESCRIPTION
**Pull request by prateek2211**
Fixes: https://tracker.ceph.com/issues/39619
Signed-off-by: Prateek Sachan psachan@cs.iitr.ac.in

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_acl.cc [hunk 1]
```diff
@@ -119,7 +119,7 @@ uint32_t RGWAccessControlPolicy::get_perm(const DoutPrefixProvider* dpp,
                                           const char * const http_referer,
                                           bool ignore_public_acls)
 {
-  ldpp_dout(dpp, 20) << "-- Getting permissions begin with perm_mask=" << perm_mask
+  ldpp_dout(dpp, 20) << "-- Getting permissions begin                                                                                with perm_mask=" << perm_mask
```
Comment (mattbenjamin) [inline src/rgw/rgw_acl.cc:122]: why this change?

Comment (prateek2211) [inline src/rgw/rgw_acl.cc:122]:
```suggestion
  ldpp_dout(dpp, 20) << "-- Getting permissions begin with perm_mask=" << perm_mask
```

Comment (prateek2211) [reply src/rgw/rgw_acl.cc]: sorry, my mistake!

**Comment by prateek2211** [comment]
I don't really know about its impact on code quality but dependence on 3rd party library is reduced for sure.

Since this being my first pr in ceph, so I chose to do cleanup sort of issues.

**Comment by prateek2211** [comment]
@mattbenjamin Also, there's problem in using std::string_view with `boost::ends_with` and `boost::starts_with` as it doesn't get converted into a boost object.
So, how about using 
```cpp
const bool starts_with(const std::string_view str, const std::string_view sub) const { 
      return (str.compare(0, sub.size(), sub)) == 0;                                     
  } 
```
as our own function

**Comment by yuvalif** [comment]
> does this significantly improve code quality? I'd love to see efforts concentrated on bugfixes and feature enhancements, tbh.



> I don't really know about its impact on code quality but dependence on 3rd party library is reduced for sure.
> Btw, this is an issue in ceph tracker https://tracker.ceph.com/issues/39619
> 
> Since this being my first pr in ceph, so I chose to do cleanup sort of issues.

also note that boost::string_ref is being deprecated by the boost team:
https://lists.boost.org/Archives/boost/2016/02/227796.php

**Comment by smithfarm** [comment]
@prateek2211 Please add a `Signed-off-by:` line to your commit messages. See https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#sign-your-work for details.

**Comment by prateek2211** [comment]
done

**Comment by mattbenjamin** [comment]
@prateek2211 that will work (ends_with);  it seems surprising that boost is deprecating string_ref, but not incorporating std::string_ref into their own interfaces in boost::algorithm::string?

**Comment by prateek2211** [comment]
yes, it worked. I was using wrong namespace

**Comment by yuvalif** [comment]
please make sure that you removed the includes as well:
```
#include <boost/utility/string_view.hpp>
#include <boost/utility/string_ref.hpp>
```

File: src/common/escape.h [hunk 2]
```diff
@@ -50,13 +49,13 @@ void escape_json_attr(const char *buf, size_t src_len, char *out);
 //   std::cout << xml_stream_escaper(xml_input) << std::endl;
 
 struct xml_stream_escaper {
-  boost::string_view str;
+  std::string_view str;
```
Comment (tchaikov) [inline src/common/escape.h:52]: please 
```c++
#include <string_view>
```
for using `std::string_view`.

Comment (tchaikov) [reply src/common/escape.h:53]: this also applies to other header files where `<boost/utility/string_view.hpp>` is removed.

Comment (prateek2211) [reply src/common/escape.h:53]: done

**Comment by yuvalif** [comment]
jenkins test docs

**Comment by prateek2211** [comment]
jenkins test docs

**Comment by prateek2211** [comment]
https://shaman.ceph.com/builds/ceph/refactor/268f75bb8d3f19eddc149ee7d963ee6537f2db31/

**Comment by prateek2211** [comment]
jenkins test docs

**Comment by yuvalif** [comment]
> https://shaman.ceph.com/builds/ceph/refactor/268f75bb8d3f19eddc149ee7d963ee6537f2db31/
> 
> http://pulpito.front.sepia.ceph.com/prateek-2020-06-19_11:11:06-rgw-refactor-distro-basic-smithi/

* there are several of these failures:
```
cluster [WRN] Health check failed: Reduced data availability: 1 pg inactive, 1 pg peering (PG_AVAILABILITY)" in cluster log
```
they unrelated to this PR, but these were already fixed. if you rebase/build and run teuthology again, you should probably not see them anymore
* multisite is failing in "test_multipart_object_sync":
```
2020-06-20T10:25:17.103 INFO:tasks.rgw_multisite_tests:======================================================================
2020-06-20T10:25:17.104 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_multipart_object_sync
2020-06-20T10:25:17.104 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2020-06-20T10:25:17.105 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2020-06-20T10:25:17.105 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
2020-06-20T10:25:17.105 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2020-06-20T10:25:17.106 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_refactor/qa/tasks/rgw_multi/tests.py", line 1200, in test_multipart_object_sync
2020-06-20T10:25:17.106 INFO:tasks.rgw_multisite_tests:    zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
2020-06-20T10:25:17.106 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_refactor/qa/tasks/rgw_multi/tests.py", line 396, in zonegroup_bucket_checkpoint
2020-06-20T10:25:17.107 INFO:tasks.rgw_multisite_tests:    zone_bucket_checkpoint(target_conn.zone, source_conn.zone, bucket_name)
2020-06-20T10:25:17.107 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_refactor/qa/tasks/rgw_multi/tests.py", line 388, in zone_bucket_checkpoint
2020-06-20T10:25:17.107 INFO:tasks.rgw_multisite_tests:    (target_zone.name, source_zone.name, bucket_name)
2020-06-20T10:25:17.108 INFO:tasks.rgw_multisite_tests:AssertionError: failed bucket checkpoint for target_zone=test-zone2 source_zone=test-zone1 bucket=ahkzgk-65
```
and also in "test_bucket_versioning":
```
2020-06-20T10:52:04.239 INFO:tasks.rgw_multisite_tests:======================================================================
2020-06-20T10:52:04.240 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_bucket_versioning
2020-06-20T10:52:04.240 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2020-06-20T10:52:04.241 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2020-06-20T10:52:04.241 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
2020-06-20T10:52:05.041 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2020-06-20T10:52:05.042 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_refactor/qa/tasks/rgw_multi/tests.py", line 822, in test_bucket_versioning
2020-06-20T10:52:05.042 INFO:tasks.rgw_multisite_tests:    assert(key in res and res[key] == 'Enabled')
2020-06-20T10:52:05.043 INFO:tasks.rgw_multisite_tests:AssertionError
```
@casey are these known issues?
* this one seems like a test issue:
```
2020-06-20T09:49:53.466 INFO:teuthology.orchestra.run.smithi050.stderr:'CommandManager' object has no attribute 'add_command_group'
2020-06-20T09:49:53.466 INFO:teuthology.orchestra.run.smithi050.stderr:Traceback (most recent call last):
2020-06-20T09:49:53.467 INFO:teuthology.orchestra.run.smithi050.stderr:  File "/home/ubuntu/cephtest/keystone/.tox/venv/lib/python3.6/site-packages/cliff/app.py", line 264, in run
2020-06-20T09:49:53.467 INFO:teuthology.orchestra.run.smithi050.stderr:    self.initialize_app(remainder)
2020-06-20T09:49:53.467 INFO:teuthology.orchestra.run.smithi050.stderr:  File "/home/ubuntu/cephtest/keystone/.tox/venv/lib/python3.6/site-packages/openstackclient/shell.py", line 133, in initialize_app
2020-06-20T09:49:53.468 INFO:teuthology.orchestra.run.smithi050.stderr:    super(OpenStackShell, self).initialize_app(argv)
2020-06-20T09:49:53.468 INFO:teuthology.orchestra.run.smithi050.stderr:  File "/home/ubuntu/cephtest/keystone/.tox/venv/lib/python3.6/site-packages/osc_lib/shell.py", line 442, in initialize_app
2020-06-20T09:49:53.468 INFO:teuthology.orchestra.run.smithi050.stderr:    self._load_plugins()
2020-06-20T09:49:53.468 INFO:teuthology.orchestra.run.smithi050.stderr:  File "/home/ubuntu/cephtest/keystone/.tox/venv/lib/python3.6/site-packages/openstackclient/shell.py", line 104, in _load_plugins
2020-06-20T09:49:53.468 INFO:teuthology.orchestra.run.smithi050.stderr:    self.command_manager.add_command_group(cmd_group)
2020-06-20T09:49:53.469 INFO:teuthology.orchestra.run.smithi050.stderr:AttributeError: 'CommandManager' object has no attribute 'add_command_group'
```
* "test_rgw_orphan_list.sh" is failing (not sure about the cause)

**Comment by yuvalif** [comment]
latest teuthology run (after rebase):

http://pulpito.front.sepia.ceph.com/yuvalif-2020-06-23_18:21:41-rgw-wip-yuval-35208-distro-basic-smithi/

all of the above test issues exist in other branches as well, e.g.
https://github.com/ceph/ceph/pull/35331#issuecomment-648661853

**Comment by cbodley** [comment]
thanks @prateek2211, nice work!
